packages/gatsby-link/src/index.tsx

```diff
@@ -1,12 +1,48 @@
 import PropTypes from "prop-types"
 import React from "react"
-import { Link } from "@reach/router"
+import {
+  Link,
+  LinkProps,
+  LinkGetProps,
+  NavigateFn,
+  NavigateOptions,
+} from "@reach/router"
 
 import { parsePath } from "./parse-path"
 
 export { parsePath }
 
-export function withPrefix(path) {
+export interface IGatsbyLinkState {
+  IOSupported: boolean
+}
+
+// eslint-disable-next-line @typescript-eslint/interface-name-prefix
+export interface GatsbyLinkProps<TState = unknown>
+  extends React.PropsWithoutRef<LinkProps<TState>> {
```

Comment:
what does `React.PropsWithoutRef` give us here?

Reply:
It gives us all of `LinkProps<TState>` except for the "ref" property.

It isn't entirely clear to me how the @reach/router `LinkProps` type ends up getting a ref type (somewhere from react's core typings), but it ends up having the property:

```
    ref: Link<TState>
```

which ends up getting confused with Gatsby's view of the ref: an HTMLAnchorElement.

I can dig more into the disconnect.

Reply:
@blainekasten  Ah, here's the problem:

the reach router LinkProps incorrectly includes a "ref" type:

```
type RefType = LinkProps<unknown>["ref"];
```

which is: `React.Ref<HTMLAnchorElement>`.

Note that in React, `ref` isn't supposed to be recorded on `props` per se, but on the actual component's intrinsic types.

But @reach/router typings define LinkProps to extend AnchorProps, which eventually include a "ref". While this is the correct ref type on @reach/router's `Link`, it shouldn't be on props.

Anyway, so **why** does "ref" existing become an issue? Because of how `GatsbyLink` is defined:

```
const GatsbyLink = React.forwardRef((props, ref) =>
  <GatsbyLinkInternal innerRef={ref} {...props} />
));
```

the problem is that `...props` is of type `GatsbyLinkProps`, which transitively includes "ref".

BUT `<GatsbyLinkInternal ref=...>` comes from `IntrinsicClassAtributes<GatsbyLinkInternal>` which is `React.Ref<GatsbyLinkInternal>`.... not at all what we want.